src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,181 @@
+/* Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2017, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+
+#ifndef NUPIC_UTIL_SLIDING_WINDOW_HPP
+#define NUPIC_UTIL_SLIDING_WINDOW_HPP
+
+#include <vector>
+#include <algorithm>
+#include <iterator>
+#include <cmath>
+#include <string>
+
+#include <nupic/types/Types.hpp>
+#include <nupic/utils/Log.hpp>
+#include <nupic/math/Math.hpp> //for macro ASSERT_INPUT_ITERATOR
+
+namespace nupic {
+  namespace util {
+
+template<class T> 
+class SlidingWindow {
+  public:
+    SlidingWindow(UInt maxCapacity, std::string id="SlidingWindow", int debug=0) : 
+      maxCapacity(maxCapacity),
+      ID(id),
+      DEBUG(debug)
+{
+      buffer_.reserve(maxCapacity);
+      idxNext_ = 0;
+} 
+
+
+    template<class IteratorT> 
+    SlidingWindow(UInt maxCapacity, IteratorT initialData_begin, 
+      IteratorT initialData_end, std::string id="SlidingWindow", int debug=0): 
+      SlidingWindow(maxCapacity, id, debug) {
+      // Assert that It obeys the STL forward iterator concept
+      ASSERT_INPUT_ITERATOR(IteratorT);
+      for(IteratorT it = initialData_begin; it != initialData_end; ++it) {
+        append(*it);
+      }
+    }
+
+    const UInt maxCapacity;
+    const std::string ID; //name of this object
+    const int DEBUG;
```

Comment:
Can you move these down next to the other member variables? You can put an extra `public:` above them so that it's clear that these are public members.

~~~c++
public: 
    const UInt maxCapacity;
    const std::string ID; //name of this object
    const int DEBUG;
private:
    std::vector<T> buffer_;
    UInt idxNext_;
~~~